### ISSUE TITLE
Type confusion in src/njs_vmcode.c found by code audit

### ISSUE THREAD

**Issue by ret2ddme** [nginx/njs#572]
Hi @xeioex 

I found there is a type confusion between  `njs_jump_off_t` and NJS_ERROR in `njs_vmcode_interpreter` and some other function.  `njs_vmcode_interpreter` has a huge swicth case table, and some probles will be produce in this scenes.

Look at the code 
https://github.com/nginx/njs/blob/404553896792b8f5f429dc8852d15784a59d8d3e/src/njs_vmcode.c#L741-L762

The main point is there is no check of ret value
```c
njs_int_t
njs_vmcode_interpreter(njs_vm_t *vm, u_char *pc, void *promise_cap,
    void *async_ctx)
{
...skip...
            case NJS_VMCODE_FUNCTION_COPY:
                fcopy = (njs_vmcode_function_copy_t *) pc;
                ret = njs_vmcode_function_copy(vm, fcopy->function,
                                               fcopy->retval);
                break;
```
you can see other case branch have check whether the ret is NJS_ERROR (equals to -1)

what's more `njs_vmcode_function_copy` can return NJS_ERROR
```c
static njs_jump_off_t
njs_vmcode_function_copy(njs_vm_t *vm, njs_value_t *value, njs_index_t retidx)
{
    njs_value_t     *retval;
    njs_function_t  *function;

    retval = njs_scope_value(vm, retidx);

    if (!njs_is_valid(retval)) {
        *retval = *value;

        function = njs_function_value_copy(vm, retval);
        if (njs_slow_path(function == NULL)) {
            return NJS_ERROR;   <-----return error
        }
    }
```

What happen if ret==-1 in `njs_vmcode_interpreter`? At the end of `njs_vmcode_interpreter`

```c
njs_int_t
njs_vmcode_interpreter(njs_vm_t *vm, u_char *pc, void *promise_cap,
    void *async_ctx)
{
...skip
        pc += ret;
...skip
```
 Finally, pc will add -1 the assign to itself as the next pc.

**BTW, This is the answer of issue https://github.com/nginx/njs/issues/571**
before that issue was reproduced, I just add some checks to avoid illegal operation and then `return NJS_ERROR`. In my memory, there are the same code(return ERROR) in same func, so I add this return NJS_ERROR. But I have not check upon func how it operates the ret value...

Finally cause issue 571, but there will have other type confusion bug like this I guess :)

**Comment by xeioex** [comment]
Hi @ret2ddme,

Thanks for the report.

Agree, this is a bug, is is not easy to trigger though, because you have to exhaust memory first.

Here is the patch:

```diff
# HG changeset patch
# Parent  dfe0eeff0568f165f9f2b3543922221886e445c6
Added error handling for NJS_VMCODE_FUNCTION_COPY instruction.

This closes #572 issue on Github.

diff --git a/src/njs_vmcode.c b/src/njs_vmcode.c
--- a/src/njs_vmcode.c
+++ b/src/njs_vmcode.c
@@ -742,6 +742,10 @@ next:
                 fcopy = (njs_vmcode_function_copy_t *) pc;
                 ret = njs_vmcode_function_copy(vm, fcopy->function,
                                                fcopy->retval);
+                if (njs_slow_path(ret != NJS_OK)) {
+                    goto error;
+                }
+
                 break;
 
             case NJS_VMCODE_FUNCTION_FRAME:
```

It seems other cases are already handled, but you may double check that.

**Comment by ret2ddme** [comment]
Good job, I will close this issue as complete.
I know what cause issue #571 ,I patch #569 add `return NJS_ERROR` in function `njs_string_offset` but it should not return it.
In `njs_string_prototype_char_code_at`
```c
static njs_int_t
njs_string_prototype_char_code_at(njs_vm_t *vm, njs_value_t *args,
    njs_uint_t nargs, njs_index_t unused)
{
...skip
    } else {
        njs_utf8_decode_init(&ctx);

        /* UTF-8 string. */
        end = string.start + string.size;
        start = njs_string_offset(string.start, end, index);   <---- call njs_string_offset, return NJS_ERROR
        code = njs_utf8_decode(&ctx, &start, end);    <------- crash here
    }
...skip
```
I'm sorry for this mistack.
